basics/state-and-lifecycle (#28)
reactjs/de.reactjs.org: content/docs/state-and-lifecycle.md

---
id: state-and-lifecycle
title: State und Lifecycle
permalink: docs/state-and-lifecycle.html
redirect_from:
  - "docs/interactivity-and-dynamic-uis.html"
prev: components-and-props.html
next: handling-events.html
---

> Diese Seite stellt das Konzept des States und Lifecycles in einer React-Komponente vor. Du kannst eine [detailiertere API Referenz für Komponenten hier finden](/docs/react-component.html).

Reviewer: ```suggestion
Diese Seite stellt das Konzept des States und Lifecycles in einer React-Komponente vor. Hier kannst du eine [detailiertere API Referenz für Komponenten](/docs/react-component.html) finden.
```

> Diese Seite stellt das Konzept des States und Lifecycles in einer React-Komponente vor. Hier kannst du eine [detailiertere API Referenz für Komponenten](/docs/react-component.html) finden.

Reviewer 2: ```suggestion
Diese Seite stellt das Konzept von States und Lifecycles in einer React-Komponente vor. Hier kannst du eine [detailiertere API Referenz für Komponenten](/docs/react-component.html) finden.
```

> Gucken wir uns das Beispiel der tickenden Uhr aus dem [vorherigen Abschnitt](/docs/rendering-elements.html#updating-the-rendered-element) an. In dem Kapitel [Elemente rendern](/docs/rendering-elements.html#rendering-an-element-into-the-dom), haben wir nur einen Weg gelernt, die UI zu aktualisieren. Wir rufen `ReactDOM.render()` auf, um die gerenderte Ausgabe zu ändern:

Reviewer: ```suggestion
Schauen wir uns das Beispiel der tickenden Uhr aus dem [vorherigen Abschnitt](/docs/rendering-elements.html#updating-the-rendered-element) an. Im Kapitel [Elemente rendern](/docs/rendering-elements.html#rendering-an-element-into-the-dom), haben wir nur einen Weg kennengelernt, das UI zu aktualisieren. Wir rufen `ReactDOM.render()` auf, um die gerenderte Ausgabe zu ändern:
```

```js{8-11}
function tick() {
  const element = (
    <div>
      <h1>Hallo Welt!</h1>

>       <h2>Es ist {new Date().toLocaleTimeString()}.</h2>

Reviewer: Should we really also translate the code examples? I did not do that in the Composition vs. Inheritance page.

Author: We looked at other translations and they did.

Reviewer: "Leave code blocks untranslated -- It is usually OK to translate comments and strings inside them." (https://github.com/reactjs/pt-BR.reactjs.org/blob/master/GLOSSARY.md)

setInterval(tick, 1000);
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/gwoJZk?editors=0010)

> In diesem Abschnitt lernen wir, wie wir die `Clock` Komponente wirklich wiederverwenbar und in sich gekapselt machen. Sie richtet sich ihren eigenen Timer ein und aktualisiert sich jede Sekunde.

Reviewer: ```suggestion
In diesem Abschnitt werden wir lernen, wie wir die `Clock` Komponente wirklich wiederverwendbar und in sich gekapselt machen. Sie richtet sich ihren eigenen Timer ein und aktualisiert sich jede Sekunde.
```

> Wir können mit der Kapslung starten, so sieht die Uhr aus:

Reviewer: ```suggestion
Wir können mit der Kapselung des Erscheinungsbildes der Uhr anfangen:
```

setInterval(tick, 1000);
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/dpdoYR?editors=0010)

> Es fehlt jedoch eine entscheidende Anforderung: Die Tatsache, dass `Clock` einen Timer aufsetzt und die UI jede Sekunden aktualisiert sollte ein Teil der Implementierung von `Clock` sein.

Reviewer: ```suggestion
Es fehlt jedoch eine entscheidende Anforderung: Die Tatsache, dass `Clock` einen Timer anlegt und das UI jede Sekunden aktualisiert sollte ein Detail der Implementierung von `Clock` sein.
```

);
```

Um dies zu implementieren, müssen wir der `Clock` Komponente einen "State" hinzufügen.

> Der State ist ähnlich wie die Props, aber er ist privat und wird vollständig von der Komponente kontrolliert.

Reviewer: ```suggestion
Der State ähnelt den Props, aber er ist privat und wird vollständig von der Komponente kontrolliert.
```

> Wir haben schon [vorher erwähnt](/docs/components-and-props.html#functional-and-class-components), dass Komponenten die als Klassen definiert werden, ein paar zusätzliche Features haben. Der lokale State genau das: Ein Feature, welches nur in Klassen verfügbar ist.

Reviewer: ```suggestion
Wir haben schon [vorher erwähnt](/docs/components-and-props.html#functional-and-class-components), dass Komponenten die als Klassen definiert werden, über einige zusätzliche Features verfügen. Der lokale State ist genau das: Ein Feature, welches nur in Klassen verfügbar ist.
```

## Umwandeln einer Funktion in eine Klasse {#converting-a-function-to-a-class}

> Du kannst in fünf Schritten eine Komponente, wie `Clock`, in eine Klasse umwandeln:

Reviewer: ```suggestion
Eine Komponente wie `Clock` kannst du in fünf Schritten in eine Klasse umwandeln:
```

> 1. Erstelle eine [ES6 Klasse](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) mit dem selben Namen, die mit Hilfe von `extends` durch `React.Component` erweitert wird.

Reviewer: ```suggestion
1. Erstelle eine [ES6 Klasse](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) mit dem selben Namen, die `React.Component` erweitert.
```

2. Erstelle eine leere Methode namens `render()`.

3. Verschiebe den Inhalt der Funktion in die `render()` Methode.

4. Ersetze `props` mit `this.props` in der `render()` Methode.

> 5. Lösche alle übrigen leeren Funktionsdeklarationen.

Reviewer: ```suggestion
5. Lösche alle übrigen, leeren Funktionsdeklarationen.
```

```js
class Clock extends React.Component {
  render() {
    return (
      <div>
        <h1>Hallo Welt!</h1>
        <h2>Es ist {this.props.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/zKRGpo?editors=0010)

`Clock` ist nun als Klasse und nicht mehr als Funktion definiert.

> Die `render` Methode wird jedes mal aufgerufen, wenn ein Update stattfindet, aber solange wir `<Clock />` in den selben DOM-Knoten rendern, wird nur eine einzige Instanz der `Clock` Klasse verwendet. Diese lässt uns zusätzliche Features, wie den lokalen State und Lifecycle Methoden verwenden.

Reviewer: ```suggestion
Die `render` Methode wird jedes mal aufgerufen, wenn ein Update stattfindet, aber solange wir `<Clock />` in den selben DOM-Knoten rendern, wird nur eine einzige Instanz der `Clock` Klasse verwendet. Diese lässt uns zusätzliche Features, wie den lokalen State und Lifecycle-Methoden verwenden.
```

Reviewer: I'm not entirely sure whether "Lifecycle-Methoden" should have the hyphen or not. But it seems like it is commonly present in similar terms.

Author: I think we should write it with a hyphen, because the two words are belong together.

## Lokalen State einer Klasse hinzufügen {#adding-local-state-to-a-class}

Wir verschieben `date` von Props hin zum State in drei Schritten:

1) Ersetze `this.props.date` mit `this.state.date` in der `render()` Methode:

```js{6}
class Clock extends React.Component {
  render() {
    return (
      <div>
        <h1>Hallo Welt!</h1>
        <h2>Es ist {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

> 2) Füge einen [Klassenkonstruktor](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes#Constructor) der den initialen `this.state` zuweist, hinzu:

Reviewer: ```suggestion
2) Füge einen [Klassenkonstruktor](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes#Constructor) hinzu, der das initiale `this.state` zuweist:
```

  render() {
    return (
      <div>
        <h1>Hallo Welt!</h1>
        <h2>Es ist {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

> Beachte wie wir `props` an den Basiskonstruktor geben:

Reviewer: ```suggestion
Beachte wie wir `props` an den Basiskonstruktor übergeben:
```

  }
```

Klassenkomponenten sollten den Basiskonstruktor immer mit `props` aufrufen.

3) Entferne das `date` Prop vom `<Clock />` Element:

```js{2}
ReactDOM.render(
  <Clock />,
  document.getElementById('root')
);
```

> Wir werden später den Code des Timers zurück in die Komponente fügen.

Reviewer: ```suggestion
Wir werden später den Code des Timers zurück in die Komponente einfügen.
```

);
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/KgQpJd?editors=0010)

Als nächstes werden wir dafür sorgen, dass `Clock` einen eigenen Timer einrichtet und sich jede Sekunde aktualisiert.

## Lifecycle Methoden zu einer Klasse hinzufügen {#adding-lifecycle-methods-to-a-class}

> In Anwendungen mit vielen Komponenten, ist es wichtig Ressourcen wieder freizugeben, wenn die Komponente gelöscht wird.

Reviewer: ```suggestion
In Anwendungen mit vielen Komponenten ist es wichtig Ressourcen wieder freizugeben, wenn die Komponente gelöscht wird.
```

> Wir wollen einen [Timer aufsetzen](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setInterval), wenn `Clock` zum ersten Mal ins DOM gerendert wird. Dies wird in React als "mounting" genannt.

Reviewer: ```suggestion
Wir wollen einen [Timer anlegen](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setInterval), wenn `Clock` zum ersten Mal ins DOM gerendert wird. Dies wird in React als "mounting" bezeichnet.
```

> Wir wollen auch, dass der [Timer entfernt](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/clearInterval) wird, wenn das von `Clock` erstellte DOM Element gelöscht wird. Dies wird in React als "unmounting" genannt.

Reviewer: ```suggestion
Wir wollen auch, dass der [Timer entfernt](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/clearInterval) wird, wenn das von `Clock` erstellte DOM Element gelöscht wird. Dies wird in React als "unmounting" bezeichnet.
```

> Wir können spezielle Methoden in der Komponentenklasse deklarieren um in diesen beim mount und unmounting etwas Code auszuführen:

Reviewer: ```suggestion
Wir können spezielle Methoden in der Komponentenklasse deklarieren um in dieser beim mounting und unmounting bestimmten Code auszuführen:
```

  render() {
    return (
      <div>
        <h1>Hallo Welt!</h1>
        <h2>Es ist {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

Diese Methoden werden "Lifecycle Methoden" genannt.

> Die `componentDidMount()` Methode wir nachdem die Komponenten in das DOM gerendert wurde ausgeführt. Dies ist eine gute Stelle um den Timer aufzusetzen:

Reviewer: ```suggestion
Die `componentDidMount()` Methode wird ausgeführt nachdem die Komponenten in das DOM gerendert wurde. Dies ist eine gute Stelle um den Timer anzulegen:
```

  }
```

Beachte, wie wir die Timer-ID in `this` speichern.

> Während `this.reps` von React selbst eingerichtet wird und `this.state` eine besondere Bedeutung hat, steht es dir frei, der Klasse manuell zusätzliche Felder hinzuzufügen, wenn du etwas was nicht am Datenfluss teilnimmt (wie eine Timer-ID) speichern musst.

Reviewer: ```suggestion
Während `this.props` von React selbst eingerichtet wird, und `this.state` eine besondere Bedeutung hat, steht es dir frei, der Klasse manuell zusätzliche Felder hinzuzufügen, wenn du etwas speichern musst, das nicht am Datenfluss teilnimmt (wie eine Timer-ID).
```

> Wir werden den Timer in der `componentWillUnmount()` Lifecycle Mthode abbrechen:

Reviewer: ```suggestion
Wir werden den Timer in der `componentWillUnmount()` Lifecycle Methode abbrechen:
```

);
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/amqdNA?editors=0010)

Jetzt tickt die Uhr jede Sekunde.

Lass uns kurz zusammenfassen, was hier vor sich geht und in welcher Reihenfolge die Methoden aufgerufen werden:

> 1) Wenn `Clock` an `ReactDOM.render()` gegeben wird, ruft React den Konstruktor der `Clock` Komponente auf. Da `Clock` die aktuelle Uhrzeit anzeigen muss, initialisert es `this.state` mit einem Objekt, welches die aktuelle Uhrzeit enthält. Wir werden diesen State später aktualisieren.

Reviewer: ```suggestion
1) Wenn `Clock` an `ReactDOM.render()` gegeben wird, ruft React den Konstruktor der `Clock` Komponente auf. Da `Clock` die aktuelle Uhrzeit anzeigen muss, initialisiert es `this.state` mit einem Objekt, welches die aktuelle Uhrzeit enthält. Wir werden diesen State später aktualisieren.
```

> 2) React ruft dann die `render()` Methode der `Clock` Komponente auf. So lernt React, was auf dem Bildschirm angezeigt werden soll. Dann aktualisiert React das DOM entsprechend der gerenderten Ausgabe von `Clock`.

Reviewer: ```suggestion
2) React ruft dann die `render()` Methode der `Clock` Komponente auf. So weiß React, was auf dem Bildschirm angezeigt werden soll. Dann aktualisiert React das DOM entsprechend der gerenderten Ausgabe von `Clock`.
```

> 3) Wenn die Ausgabe von `Clock` in das DOM eingefügt wurde, ruft React die Lifecycle Methode `componentDidMount()` auf. In dieser fordert die `Clock` Komponente den Browser auf einen Timer aufzusetzen, der jede Sekunde die `tick()` Methode aufruft.

Reviewer: ```suggestion
3) Wenn die Ausgabe von `Clock` in das DOM eingefügt wurde, ruft React die Lifecycle Methode `componentDidMount()` auf. In dieser fordert die `Clock` Komponente den Browser auf einen Timer anzulegen, der jede Sekunde die `tick()` Methode aufruft.
```

> 4) Jede Sekunde ruft der Browser die `tick()` Methode auf. Innerhalb dieser plant die `Clock` Komponente das Aktualisieren der UI in dem `setState()` mit einem Objekt, welches die aktuelle Uhrzeit beinhaltet, aufruft. Dank des `setState()` aufrufs, weiß React, dass sich der State geändert hat und ruft die `render()` Methode auf um erneut zu erfahren, was auf dem Bildschirm dargestellt werden soll. Diesmal wird `this.state.date` in der `render()` Methode anders sein und der gerenderte Inhalt die aktualsierte Zeit beinhalten. React aktualisiert dementsprechend das DOM.

Reviewer 2: ```suggestion
4) Jede Sekunde ruft der Browser die `tick()` Methode auf. Innerhalb dieser plant die `Clock` Komponente das Aktualisieren der UI in dem `setState()` mit einem Objekt, welches die aktuelle Uhrzeit beinhaltet, aufruft. Dank des `setState()`-Aufrufs weiß React, dass sich der State geändert hat und ruft die `render()` Methode auf um erneut zu erfahren, was auf dem Bildschirm dargestellt werden soll. Diesmal wird `this.state.date` in der `render()` Methode anders sein und der gerenderte Inhalt die aktualsierte Zeit beinhalten. React aktualisiert dementsprechend das DOM.
```

5) Wenn die `Clock` Komponente jemals aus dem DOM entfernt wird, ruft React die `componentWillUnmount()` Lifecycle Methode auf und der Timer stoppt.

> ## Benutze State korrekt {#using-state-correctly}

Reviewer: ```suggestion
## State korrekt verwenden {#using-state-correctly}
```

> Es gibt drei Dinge die du über `setState()` wissen solltest.

Reviewer 2: ```suggestion
Es gibt drei Dinge, die du über `setState()` wissen solltest.
```

> ### Bearbeite den State nicht direkt {#do-not-modify-state-directly}

Reviewer: ```suggestion
### Bearbeite State nicht direkt {#do-not-modify-state-directly}
```

> Dies wird zum Beispiel kein neu Rendern der Komponente bewirken:

Reviewer 2: ```suggestion
Dies wird zum Beispiel kein neurendern der Komponente bewirken:
```

```js
// Wrong
this.state.comment = 'Hello';
```

Benutze stattdessen `setState()`:

```js
// Correct
this.setState({comment: 'Hello'});
```

Der einzige Ort, an dem du `this.state` setzen kannst, ist der Konstruktor.

### Stateaktualsierungen können asynchron sein {#state-updates-may-be-asynchronous}

React kann aus Preformancegründen mehrere `setState()` Aufrufe in einem Update zusammenfassen.

> Da `this.props` und `this.state` asynchron aktualisiert werden können, solltest du nicht darauf verlassen, die Werte für Berechnungen des nächsten State verwenden zu können.

Reviewer: ```suggestion
Da `this.props` und `this.state` asynchron aktualisiert werden können, solltest du dich nicht darauf verlassen, die Werte für Berechnungen des nächsten State verwenden zu können.
```

});
```

> Um dies zu verhindern, benutzen wir eine zweite Form von `setState()`, die eine Funktion anstatt eines Objektes entgegennimmt. Diese Funktion besitzt als erstes Argument den vorherigen State und die Props zum Zeitpunkt der Aktualisierung, als zweites Argument:

Reviewer: ```suggestion
Um dies zu verhindern, verwenden wir eine zweite Form von `setState()`, die eine Funktion anstatt eines Objektes entgegennimmt. Diese Funktion besitzt als erstes Argument den vorherigen State und die Props zum Zeitpunkt der Aktualisierung als zweites Argument:
```

}));
```

> Wir haben eine [Lamda Funktion](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) oben verwendet, aber normale Funktionen funktionieren ebenso:

Reviewer: ```suggestion
Wir haben oben eine [Lamda Funktion](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) verwendet, aber normale Funktionen funktionieren ebenso:
```

});
```

> ### Stateaktualsierungen werden zusammengeführt {#state-updates-are-merged}

Reviewer 2: ```suggestion
### Stateaktualisierungen werden zusammengeführt {#state-updates-are-merged}
```

  }
```

Das Zusammenführen ist nur oberflächlich und `this.setState({comments})` lässt `this.state.posts` bestehen, aber ersetzt `this.state.comments`.

## Der "top-down" Datenfluss {#the-data-flows-down}

> Weder parent noch child Komponenten können wissen ob eine bestimmte Komponente stateful oder stateless ist, uns sollten sich auch nicht darum kümmern ob sie als Funktion oder Klasse definiert wurde.

Reviewer: ```suggestion
Weder Eltern- noch Kind-Komponenten können wissen ob eine bestimmte Komponente stateful oder stateless ist, und sollten sich auch nicht darum kümmern ob sie als Funktion oder Klasse definiert wurde.
```

Deshalb wird der State oft als lokal oder gekapselt bezeichnet. Er ist nur für die Komponente, die ihn besitzt und erstellt hat zugänglich. Für sonst keine andere Komponente.

Eine Komponente kann sich aussuchen ob sie ihren State als Props weitergeben möchte:

```js
<h2>Es ist {this.state.date.toLocaleTimeString()}.</h2>
```

Dies funktioniert auch für benutzerdefinierte Komponenten:

```js
<FormattedDate date={this.state.date} />
```

> Die `FormattedDate` Komponente nimmt `date` als Prop entgegen und wüsste nicht ob es aus dem `Clock`'s State, Props oder von Handeingabe kommt:

Reviewer: ```suggestion
Die `FormattedDate` Komponente nimmt `date` als Prop entgegen und wüsste nicht ob es aus dem State oder den Props von `Clock` stammt, oder von Hand eingegeben wurde:
```

);
```

[**Probier es auf CodePen aus**](https://codepen.io/gaearon/pen/vXdGmd?editors=0010)

> Jede `Clock` setzt ihren eigenen Timer auf und aktualisiert sich selbstständig.

Reviewer: ```suggestion
Jede `Clock` legt ihren eigenen Timer an und aktualisiert sich selbstständig.
```

> In React Apps wird das Implementierungsdetail der Komponente, das sich im Laufe der Zeit ändern kann, als stateless oder stateful betrachtet. Du kannst stateless Komponenten innerhalb svon tateful Komponenten benutzen und umgekehrt.

Reviewer: ```suggestion
In React Apps wird das Implementierungsdetail der Komponente, das sich im Laufe der Zeit ändern kann, als stateless oder stateful betrachtet. Du kannst stateless Komponenten innerhalb von stateful Komponenten benutzen und umgekehrt.
```

Reviewer: Another case where I'm not sure whether terms should be translated or not (stateful/stateless). "Zustandslos" sounds fine to me, but  "Zustandshaltend" or "Zustandsbehaftet" a bit odd...

Author: I was also thinking about it. We should discuss it in slack.
